File: imagegen/imagegen-cli/src/lib/api-key.ts
/**
 * Resolve the Gemini API key.
 *
 * Order:
 *   1. `GEMINI_API_KEY` environment variable (wins; supports CI / explicit override).
 *   2. `.gemini-key` file walked up from `cwd` to filesystem root (project-local).
 *   3. `~/.gemini-key` in the user's home directory (machine-wide default).
 *
 * The file must contain just the raw key (whitespace is trimmed). Users should
 * add `.gemini-key` to `.gitignore` since it holds a secret.
 */

import { readFile, stat } from "node:fs/promises";
import { homedir } from "node:os";
import { dirname, resolve } from "node:path";

export type ApiKeySource = "env" | "file";

export interface ResolvedApiKey {
	key: string;
	source: ApiKeySource;
	/** Absolute path to the file when source === "file". */
	path?: string;
}

const KEY_FILENAME = ".gemini-key";

async function fileExists(p: string): Promise<boolean> {
	try {
		const s = await stat(p);
		return s.isFile();
	} catch {
		return false;
	}
}

async function findProjectKeyFile(startDir: string): Promise<string | null> {
	let dir = resolve(startDir);
	// Cap iterations defensively; we stop when dirname returns the same path (filesystem root).
	for (let i = 0; i < 64; i++) {
		const candidate = resolve(dir, KEY_FILENAME);
		if (await fileExists(candidate)) return candidate;
		const parent = dirname(dir);
		if (parent === dir) return null;
		dir = parent;
	}
	return null;
}

async function readKeyFile(path: string): Promise<string | null> {
	const raw = await readFile(path, "utf8");
	const key = raw.trim();
	return key || null;
}

export async function resolveApiKey(
	cwd: string = process.cwd(),
): Promise<ResolvedApiKey | null> {
	const fromEnv = process.env.GEMINI_API_KEY?.trim();
	if (fromEnv) {
		return { key: fromEnv, source: "env" };
	}

	const projectPath = await findProjectKeyFile(cwd);
	if (projectPath) {
		const key = await readKeyFile(projectPath);
		if (key) return { key, source: "file", path: projectPath };
	}

	const home = homedir();
	if (home) {
		const homePath = resolve(home, KEY_FILENAME);
		if (await fileExists(homePath)) {
			const key = await readKeyFile(homePath);
			if (key) return { key, source: "file", path: homePath };
		}
	}

	return null;
}

File: imagegen/imagegen-cli/src/imagegen_cli/api_key.py
"""Resolve the Gemini API key.

Order:
  1. ``GEMINI_API_KEY`` environment variable (wins; supports CI / explicit override).
  2. ``.gemini-key`` file walked up from ``cwd`` to filesystem root (project-local).
  3. ``~/.gemini-key`` in the user's home directory (machine-wide default).

The file must contain just the raw key (whitespace is trimmed). Users should
add ``.gemini-key`` to ``.gitignore`` since it holds a secret.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path

KEY_FILENAME = ".gemini-key"


@dataclass(frozen=True)
class ResolvedApiKey:
    key: str
    source: str  # "env" | "file"
    path: Path | None = None


def _find_project_key_file(start: Path) -> Path | None:
    cur = start.resolve()
    # Defensive cap; we stop when parent == cur (filesystem root).
    for _ in range(64):
        candidate = cur / KEY_FILENAME
        if candidate.is_file():
            return candidate
        parent = cur.parent
        if parent == cur:
            return None
        cur = parent
    return None


def _read_key_file(path: Path) -> str | None:
    raw = path.read_text(encoding="utf-8")
    key = raw.strip()
    return key or None


def resolve_api_key(cwd: Path | None = None) -> ResolvedApiKey | None:
    env = (os.environ.get("GEMINI_API_KEY") or "").strip()
    if env:
        return ResolvedApiKey(key=env, source="env")

    project = _find_project_key_file(cwd or Path.cwd())
    if project is not None:
        key = _read_key_file(project)
        if key:
            return ResolvedApiKey(key=key, source="file", path=project)

    home_path = Path.home() / KEY_FILENAME
    if home_path.is_file():
        key = _read_key_file(home_path)
        if key:
            return ResolvedApiKey(key=key, source="file", path=home_path)

    return None
